### PR TITLE
Week 4

### PR DESCRIPTION
**Pull request by AimeeBorda**
both approaches are implemented - flood folder contains the DFS for each island approach and unionFind folder contains the approach using a simple implementation of unionFind (give the structure of this problem, the groups are flat rather than as a tree simplifying most of the other operations)

**changes from the feedback have been implemented**

### CODE REVIEW

File: src/main/java/week4/unionFind/IslandCountUF.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            count--;
+          } else if (c < cols - 1
+              && map[r][c + 1]
+              && uf[getIndex(r, c + 1, rows)] != uf[getIndex(r, c, rows)]) {
+            count--;
+            merge(uf, getIndex(r, c, rows), getIndex(r, c + 1, rows));
+          }
+        }
+      }
+    }
```
Comment (simon-frankau) [inline src/main/java/week4/unionFind/IslandCountUF.java]: I _think_ this doesn't work to fix it up, with sufficiently complicated inputs. However, I'm going to be slightly cheeky and say "Even if this code isn't obviously incorrect, it's not obviously correct, either. Code review shouldn't be hard work, and this code should be refactored/commented until it's clear that it works as intended."

File: src/main/java/week4/IslandCountTest.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  @Test
+  public void testIslandWithHole() {
+    boolean[][] map = new boolean[][]{
+        {false, true, true, true},
+        {false, true, false, true},
+        {false, true, true, true},
+        {false, false, false, false}
+    };
+    assertEquals(1, islandCount.countIslands(map));
+  }
```
Comment (simon-frankau) [inline src/main/java/week4/IslandCountTest.java]: Ah, great. I was hoping someone would think of this test case.

File: src/main/java/week4/IslandCountTest.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    map = new boolean[][]{
+        {false},
+        {false},
+        {true},
+        {false},
+        {true}
+    };
+    assertEquals(2, islandCount.countIslands(map));
+  }
+}
```
Comment (simon-frankau) [inline src/main/java/week4/IslandCountTest.java]: Looks like a good set of tests, but I still have a nasty suspicion that the union find version has bugs for complex-enough cases.

File: src/main/java/week4/flood/IslandCount.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+import java.util.HashSet;
+import java.util.Set;
+import week4.IslandCountI;
+
+public class IslandCount implements IslandCountI {
+
+  /*
+   * We go through each cell, if it is set (it is an island),
+   * we increase the count and reset all the tiles in that island
```
Comment (AimeeBorda) [inline src/main/java/week4/flood/IslandCount.java]: It should be fairly easy to implement the same solution without modifying the map array

File: src/main/java/week4/flood/IslandCount.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+          grayOutIsland(map, r, c);
+        }
+      }
+    }
+    return count;
+  }
+
+  /* sets all adjacent tiles of an island to false */
+  private void grayOutIsland(boolean[][] map, int r, int c) {
+    Set<Tile> tiles = new HashSet<>();
```
Comment (AimeeBorda) [inline src/main/java/week4/flood/IslandCount.java]: yeah that makes much more sense

File: src/main/java/week4/flood/IslandCount.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    int[] dx = new int[]{0, 0, -1, 1};
+    int[] dy = new int[]{-1, 1, 0, 0};
```
Comment (AimeeBorda) [inline src/main/java/week4/flood/IslandCount.java]: yeah - I thought so as well cause this is not readable

File: src/main/java/week4/flood/IslandCount.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    }
+  }
+
+  private Tile next(Set<Tile> tiles) {
+    Tile next = tiles.iterator().next();
+    tiles.remove(next);
+    return next;
+  }
+
+  private boolean getAndReset(boolean[][] map, Tile next) {
```
Comment (AimeeBorda) [inline src/main/java/week4/flood/IslandCount.java]: made redundant

File: src/main/java/week4/unionFind/IslandCountUF.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    int[] uf = new int[rows * cols];
+
+    Arrays.setAll(uf, i -> i);
+
+    for (int r = 0; r < rows; r++) {
+      for (int c = 0; c < cols; c++) {
+        if (map[r][c]) {
+          if (r > 0 && map[r - 1][c]) {
+            merge(uf, getIndex(r, c, rows), getIndex(r - 1, c, rows));
+          } else if (c > 0 && map[r][c - 1]) {
```
Comment (AimeeBorda) [inline src/main/java/week4/unionFind/IslandCountUF.java]: yep - found the counter example it shows up in the bottom right corner so apparently I did not test for the corners... fix still to be done

File: src/main/java/week4/IslandCountTest.java (hunk 3 above)
Comment (AimeeBorda) [inline src/main/java/week4/IslandCountTest.java]: you are right but the bug shows up with a very simple example 
.  x
x x

File: src/main/java/week4/unionFind/IslandCountUF.java (hunk 8 above)
Comment (AimeeBorda) [inline src/main/java/week4/unionFind/IslandCountUF.java]: Managed to find the counterexample which makes sense - will try and fix the problem over the weekend

File: src/main/java/week4/unionFind/IslandCountUF.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+            merge(uf, getIndex(r, c, rows), getIndex(r, c - 1, rows));
+          } else {
+            count++;
+          }
+        }
+      }
+    }
```
Comment (AimeeBorda) [inline src/main/java/week4/unionFind/IslandCountUF.java]: and also symmetry in the solution - I think?

**Pull request merged**